clients/lsp-rust.el

```diff
@@ -928,5 +937,64 @@ them with `crate` or the crate name they refer to."
   (interactive)
   (lsp-find-locations "experimental/parentModule" nil :display-action display-action))
 
+(defun lsp-rust-analyzer-open-cargo-toml (&optional new-window)
+  "Open the closest Cargo.toml from the current file.
+
+   Rust-Analyzer LSP protocol documented here and added in November 2020
+   https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#open-cargotoml
+
+   If NEW-WINDOW (interactively the prefix argument) is non-nil,
+   open in a new window."
+  (interactive "P")
+  (-if-let (workspace (lsp-find-workspace 'rust-analyzer))
+      (-if-let* ((params (lsp-make-rust-analyzer-syntax-tree-params
+                          :text-document (lsp--text-document-identifier)))
+                 ;; REVIEW is there a convention how to define and call rust-analyzer
+                 ;; lsp extension methods?
+                 (cargo-toml-location (with-lsp-workspace workspace
+                                        (lsp-send-request (lsp-make-request
+                                                           "experimental/openCargoToml"
+                                                           params))))
+                 ;; REVIEW is there a more idiomatic to retrieve the :uri key
+                 ;; or default to nil?
+                 (cargo-toml-uri (gethash "uri" cargo-toml-location)))
+          (funcall (if new-window #'find-file-other-window #'find-file)
+                   (lsp--uri-to-path cargo-toml-uri))
+        ;; REVIEW make it an error and/or add more actionable/debuggable info?
+        (lsp--warn "Couldn't find a Cargo.toml file or your version of rust-analyzer doesn't support this extension"))
+    (lsp--warn "Not running rust-analyzer")))
+
+
+(defun lsp-rust-analyzer--related-tests ()
+  "Get runnable test items related to the current TextDocumentPosition.
+Calls a rust-analyzer LSP extension endpoint that returns a wrapper over Runnable[]"
+  (lsp-send-request (lsp-make-request
+                     "rust-analyzer/relatedTests"
+                     (lsp--text-document-position-params))))
+
+(defun lsp-rust-analyzer--select-related-test ()
+  "Call the endpoint and ask for user selection.
+
+Cannot reuse `lsp-rust-analyzer--select-runnable' because the runnables endpoint
+responds with Runnable[], while relatedTests responds with TestInfo[], which is a wrapper
+over runnable. Also, this method doesn't set the `lsp-rust-analyzer--last-runnable' variable"
+    (lsp--completing-read
+     "Select test: "
+     ;; since the endpoint returns a list of hash tables that
+     ;; stores &rust-analyzer:Runnable as a value under key
+     ;; "runnable"
+     ;; we need to unpack them first
+     (mapcar (lambda (runnable) (gethash "runnable" runnable)) (lsp-rust-analyzer--related-tests))
```

Comment:
Same here, using `gethash` is discouraged for server response, since we can switch to use `plist` instead of `hashtable`

Reply:
Thanks, I am quite confused why the RelatedTests endpoint doesn't conform to the `Runnable[]` interface, I have asked the author this question (and a suggestion to reuse the interface) and will action the outcome.